examples/internal/testutil/testutil.go

```diff
@@ -0,0 +1,157 @@
+	pulumi_testing "github.com/pulumi/pulumi/sdk/v3/go/common/testing"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/retry"
+)
+
+// Test that upgrading the binary and language SDK is safe and results in no-change preview and update.
+//
+// providerBinary: name of the binary that may be in PATH, such as "pulumi-resource-random".
+//
+// Ideally this should be expressed in ProgramTest facilities, but it is currently not possible.
+//
+// Also, currently this only works for Node Pulumi programs.
+func ProviderUpdateTest(t *testing.T, providerBinary string, opts integration.ProgramTestOptions) {
+	opts.SkipEmptyPreviewUpdate = true
+	opts.SkipExportImport = true
+
+	// The environment may have providerBinary in PATH, but the first update should use a different version.
+	// Restrict the PATH so pulumi CLI does not see it.
+	opts = opts.With(restrictPath(t, "node", "pulumi", "yarn"))
+
+	// Capture workdir from PrepareProject, as integration.ProgramTester does not expose it.
+	var workdir string
+
+	// This will be called only once. Call yarn install. Assume package.json refs a prod dependency like this:
+	//
+	//    "@pulumi/random": "^4.0.0"
+	//
+	// Then this should pull the latest prod version of the provider Node SDK, and then Pulumi will auto-install the
+	// matching provider binary version.
+	opts.PrepareProject = func(info *engine.Projinfo) (err error) {
+		workdir = info.Root
+		yarn(t, opts, workdir, "install")
+		return nil
+	}
+
+	pt := integration.ProgramTestManualLifeCycle(t, &opts)
+	err := pt.TestLifeCyclePrepare()
+	require.NoError(t, err)
+
+	pt.TestFinished = false
+	defer pt.TestCleanUp()
+
+	err = pt.TestLifeCycleInitialize()
+	require.NoError(t, err)
+
+	defer func() {
+		destroyErr := pt.TestLifeCycleDestroy()
+		require.NoError(t, destroyErr)
+	}()
+
+	// This should preview and update the stack using the stable version of the provider and its SDK.
+	err = pt.TestPreviewUpdateAndEdits()
+	require.NoError(t, err)
+
+	// Now bring the local binary being tested back in PATH.
+	opts = opts.With(restrictPath(t, "node", "pulumi", "yarn", providerBinary))
+
+	// Also bring local Node SDK versions in for testing via yarn link.
+	for _, d := range opts.Dependencies {
+		yarn(t, opts, workdir, "link", d)
+	}
+
+	// Test that local provider and SDK versions generate nop preview/udpate on the stack.
+	err = pt.PreviewAndUpdate(workdir, "upgraded",
+		false, /* shouldFail */
+		true,  /* expectNopPreview */
+		true /* expectNopUpdate */)
+	require.NoError(t, err)
+
+	pt.TestFinished = true
+}
+
+func yarn(t *testing.T, opts integration.ProgramTestOptions, wd string, args ...string) {
+	name := "yarn " + strings.Join(args, " ")
+
+	// Yarn will time out if multiple processes are trying to install packages at the same time.
+	pulumi_testing.YarnInstallMutex.Lock()
+	defer pulumi_testing.YarnInstallMutex.Unlock()
+	t.Log("acquired yarn install lock")
+	t.Log("released yarn install lock")
```

Comment:
This will have log messages corresponding to when the lock is really acquired and released.
```suggestion
	pulumi_testing.YarnInstallMutex.Lock()
	defer t.Log("released yarn install lock")
	defer pulumi_testing.YarnInstallMutex.Unlock()
	t.Log("acquired yarn install lock")
```

Shouldn't `pulumi_testing` be `pulumiTesting`? I think we do camelCase for all go.

Reply:
Unabashed copy-paste from https://github.com/pulumi/pulumi/blob/master/pkg/testing/integration/program.go#L49